### PR TITLE
Road tasks

### PR DESCRIPTION
**Pull request by dereklieu**
Interface for the tasking manager

screen shot 2017-10-02 at 7 08 19 pm

### Current TODO's:

- [x] Having entered a mode (either merge or join) clicks on a road should add that road's `_id` property to the internal state. `_id` properties are randomly assigned from the conflation-detection script.
- [x] Implement merge and join (at intersection) methods. Merge should take the average or centerline geometry for all selected lines. Join should find the intersection between two center lines and create a node at that intersection, and add it to each line.
- [x] Create OSM changeset xml payloads from the above, that we can upload directly to the api.
- [x] Sync the current zoom level with the global zoom level, so you can easily switch back and forth to/from the editor.
- [ ] Maybe save the current task into localStorage, so if you do switch to a different page and come back, you don't lose the task you're working on?

### Caveats:

I'm serving data to this locally from https://github.com/dereklieu/orma-vietnam-conflation.

The action creators in this PR reflect that, and must be changed once I can get this running on our deployment, hopefully tomorrow.

To recreate my setup, open a terminal window, install the repo dependencies (`yarn`) and then run `node index.js`. That will generate the conflation data and make it available via express server.

The API also has no notion of a 'in-progress' endpoint. This is a TODO on my end for the deployment.

### Blockers:

- [ ] The data dump I'm using doesn't include Macrocosm IDs, which makes it impossible to generate a 'real' changeset targeting those ways (however we can still write the logic to make that changeset, and then add in the ID's once we are running on real data).

@danielfdsilva once the conflation script is hosted, I think it should cover a lot of our use cases for error checking. The one remaining item that my script wouldn't touch is the case of a road that *should* connect to another road but doesn't, because the field data run ended early. If you wanted to research how to tackle that, I think it would be super helpful. However, if you get stuck on that, would you be open to tackling some of the TODO's I listed here?

### CODE REVIEW

File: app/assets/scripts/views/tasks.js [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+  filter: ['all', ['!has', 'main'], ['==', '_id', '']]
+}].map(layer => Object.assign({source}, layer));
+
+const layerIds = layers.map(layer => layer.id);
+
+var Tasks = React.createClass({
+  getInitialState: function () {
+    return {
+      currentTaskId: null,
+      currentTask: null,
```
Comment (danielfdsilva) [inline app/assets/scripts/views/tasks.js]: It's a bit redundant to store both the task and the id. Can't we just store the task?

File: app/assets/scripts/views/tasks.js [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+  setNewTask: function (task) {
+    this.setState({
+      currentTaskId: task._id,
+      currentTask: task
+    });
+  },
+
+  onMapLoaded: function (fn) {
+    if (this.map.loaded()) fn();
+    else this.map.once('load', fn);
```
Comment (danielfdsilva) [inline app/assets/scripts/views/tasks.js]: A bit confusing to read perhaps?

**Comment by danielfdsilva** [comment]
Been looking at this and thinking how we could check for lines that should intersect. Honestly I've no idea what to search for but I thought that maybe we could adapt the script to look for that as well.

Currently the script searches for roads that are very close. The roads that should intersect are basically roads where the last point is close to another road. So we would just need to check whether the last point of any road is close enough to another. Would this work?

**Comment by dereklieu** [comment]
Right now linematching is done on segments (defined as straight lines of any length) that actually match. The segments are indexed by their bounding box. I think it might be feasible to modify the script so that it pads out the bounding box, then looks at the beginning and ending segments of each line rather than segments from the whole line.

However, I'm still holding out hope that there's a library out there that accomplishes this more out of the box. @matthewhanson anything you know of?

**Comment by dereklieu** [comment]
@mileswwatkins I think this is ready to go. Do you want to give it a look over to make sure everything works?

There are a few to-do's and gotchas to be aware of.

1. Right now, a single way can be included in multiple tasks. This is because the script doesn't keep track of what way has been marked as a task when it processes them. It simply looks at every way and figures out the potential suspects nearby. So, even if you skip, you might still be presented with a task that looks the same. I think there's some relatively easy id-deduping we can do here.
1. The "Finish" button marks that task in the database as pending. This lets you not see that particular task again (until the script runs again) even if you refresh. However, this doesn't work that well because of the above.
1. When joining ways together that don't currently intersect, it's important to pay attention to the way that you select first. **The join will only happen from either the beginning or end of that way.** It will join to whatever's closest of the second way you select. This is mainly for performance reasons (rather than finding the 2 closest points of 2 ways, which can be costly). It's a little confusing, but with the proper user training I think we can mitigate the effects.

**Comment by mileswwatkins** [comment]
Mr Lieu, it is always a pleasure seeing the UX that you build.

**Pull request merged**